### DISCUSSION TITLE
Seeking Greater Than 115200bps SerialUPDI (Long, Lots of Pictures)

### DISCUSSION THREAD

**Discussion by jvasileff** [SpenceKonde/megaTinyCore#595] [Q&A]
I've been struggling to achieve better than 115200bps using [SerialUPDI](https://github.com/SpenceKonde/AVR-Guidance/blob/master/UPDI/jtag2updi.md) with several adapters for a while, but was surprised recently in the other direction.  A [CH340E](https://www.mpja.com/download/35227cpdata.pdf) chip on a custom board worked at 921600bps with pretty much *any* diode and resistor configuration.

Obviously, with the CH340E the UPDI pin looked good on the scope, but the others looked fine too after adjusting resistor and diode configurations despite maxing out at 115200bps. So, I've attempted to isolate the many variables that can affect performance.

Given a capable UART adapter, I've identified the following factors that may affect SerialUPDI performance:

- Software
    - OS UART adapter driver quality and timing characteristics
    - SerialUPDI programmer (`prog.py`) quality and timing characteristics
- Hardware
    - UART adapter pin capacitance and logic levels
    - Internal RX pullup resistor
    - External series TX resistor
    - SerialUPDI diode configuration/selection
    - SerialUPDI resistor configuration/selection

Below, I'll describe my attempt to abstract away differences in serial adapter hardware through the use of buffers to a) analyze the software (driver and `prog.py`) for UART chips independently of the above hardware factors, and b) explore the limits of the UPDI hardware on ATtinys in a controlled environment.

# The Test Circuit

A [74LVC2G125](https://www.ti.com/lit/ds/symlink/sn74lvc2g125.pdf) buffer was used between the serial adapters and UPDI pin with the expectations that:

1. the buffers would be able to perform I/O at maximum possible speed on each of the serial adapters,

2. the UPDI side of the buffers would be consistent and well understood (i.e. consisent drive strength, no series resistor on TX unless one is added, etc.), and

3. for any given serial adapter, the test framework would provide the best chance of success at high bitrates.

The following test circuit was used:

![test-circuit](https://user-images.githubusercontent.com/236050/148161899-f6b46ba1-3f07-45b8-921a-6ce688fae570.jpg)

The following rationale was used for component selection:

- **U1: [74LVC2G125 Buffer](https://www.ti.com/lit/ds/symlink/sn74lvc2g125.pdf)**: This is the part I had on hand. A buffer with Schmitt trigger inputs might be better.
- **R1: 1kΩ**: This "series resistor" is for protection of the TX pin and its value primarily affects the fall time of the signal on the UPDI pin. A lower value could be used, but fall times are plenty fast with 1k (well under 1µs).
- **R2: 4.7kΩ**: This "parallel resistor" helps to speed up the signal rise time at the UPDI pin which needs to be under (or at least close to) 1µs for 921600bps communication. Higher values are discussed in more detail below.
- **D1: [1N4148 Small Signal Diode](https://www.onsemi.com/pdf/datasheet/1n914-d.pdf)**: A fast and simple diode is all that is needed. Fall times are plenty fast, and the UPDI pin does not need a super low input low-voltage value (VIL2 = 0.3 * 3.3V = 1V). A Schottky diode would add complexity in terms of speed and leakage current which is contrary to the goal of keeping the test circuit simple as possible. More on that below.

# Test Circuit Performance

The test circuit was measured to see if it met the following performance goals:

- Near identical performance at the UPDI pin for all serial adapters
- Sub-1µs rise and fall times to facilitate 921600bps programming speeds
- Adequate high and low voltage levels (low < 1V, high > 2.3V)

There are two sets of rise and fall times to consider. The first (A) is when the adapter's TX pin pulls the UPDI signal high or low. The second (B) is when the adapter's TX is held high, and the UPDI pin asserts high or low.

The following adapters were tested:

1. A fake PL2303 based USB adapter from Ali Express
2. A ~FTDI~ CP2102 USB adapter
3. A CP2105 USB adapter, using the "enhanced" port
4. A CH340E USB adapter

As the charts show below, all criteria was met. The horizontal scale is 1µs/div. The UPDI pin is in yellow, and the buffered TX in purple.

### fake-PL2303:
![fake-PL2303-74x125-1k-sig-4 7k](https://user-images.githubusercontent.com/236050/148160267-a27d1469-aaad-40de-8e55-693bf550f5c1.png)

### CP2102:
![ftdi-74x125-1k-sig-4 7k](https://user-images.githubusercontent.com/236050/148160288-e48dc0e3-d810-494b-81c8-3924c69813d9.png)

### CP2105:
![cp2105-74x125-1k-sig-4 7k](https://user-images.githubusercontent.com/236050/148160302-cfe25e9d-7b8e-402c-90bb-a298216324b3.png)

### CH340E:
![ch340e-74x125-1k-sig-4 7k](https://user-images.githubusercontent.com/236050/148160333-de663809-bafd-452d-9ae4-0c45c95c00db.png)

# So... Does it Work? What About the Drivers and Software?

As mentioned, I previously had trouble achieving speeds above 115200 with some adapters.  This appears to have been primarily a driver & software problem.

While working on this, I discovered the SerialUPDI (`prog.py`) option `-wd WRITEDELAY`:

```
    -wd WRITEDELAY, --writedelay WRITEDELAY
            Page write delay [ms] for tinyAVR and megaAVR. USB latency
            is usually sufficient without this. (Default: 0 - this
            severely impacts write performance)
```

Setting `-wd 2` improved high speed capability in many scenarios. I tested on Linux (Raspberry Pi 5.10.63-v7l+, armv7l) and MacOS (iMac Pro, Monterey 12.1) with various settings and results. All tests were performed with megaTinyCore version 2.5.4 from command line.

### Results:

Adapter | OS | Write Delay | Top Working Speed
------- | -- | ------------------- | -----------------
CP2105 | MacOS | `-wd 0` | 115200 ✗
CP2105 | MacOS | `-wd 2` | 921600 ✓
CP2105 | Linux | `-wd 0` | 230400
CP2105 | Linux | `-wd 2` | 921600 ✓
fake PL2303 | MacOS | `-wd 0` | 115200 ✗
fake PL2303 | MacOS | `-wd 2` | 230400
fake PL2303 | Linux | `-wd 0` | 115200 ✗
fake PL2303 | Linux | `-wd 2` | 921600 ✓
CH340E | MacOS | `-wd 0` | 921600 ✓
CH340E | Linux | `-wd 0` | 921600 ✓
CP2102 | MacOS | `-wd 0` | 115200 ✗
CP2102 | MacOS | `-wd 2` | 921600 ✓
CP2102 | Linux | `-wd 0` | 230400
CP2102 | Linux | `-wd 2` | 921600 ✓

So, the results are encouraging. With the right driver and software, 921600bps is quite achievable, but `prog.py` doesn't always work with default settings.

# What Else Can We Learn?

With the test framework in place, components can be adjusted to push the limits of the UPDI port which will help us better design UPDI adapters and set a reference for what *should* work from a hardware perspective.

Below I'll take a look at the need for the R2 "parallel" pullup resistor and Schottky vs. Small Signal Diode selection.

# Is the "Parallel" Pullup Resistor R2 Necessary?

A pullup resistor on the UPDI line of *some* sort is necessary to reduce rise times for high speed programming. This fact is masked in some real-world UPDI programmers as USB to UART adapters often have pullups on the RX pin, which is directly attached to the UPDI line. But unless the pullup is very strong, another pullup will be required for 921600bps.

The benefit of using a resistor parallel to the D1 diode rather than a pullup to VCC is that the pullup is only required when the TX line is high. When TX is low, the UPDI line will be pulled low, and a pullup to VCC would be counter
productive.

The test circuit calls for a `4.7k` parallel resistor, but additional values were tested to show the impact on UPDI rise times.

For short (1µs) pulses, a 22k pullup is not adequate to fully pull the UPDI line high:

![ch340e-74x125-1k-sig-22k](https://user-images.githubusercontent.com/236050/148160445-5fd501a9-669a-47be-8a79-42f5504bffc2.png)

A 33k pullup is *slightly* worse:

![ch340e-74x125-1k-sig-33k](https://user-images.githubusercontent.com/236050/148160458-e88babb2-da9f-41eb-a7a3-1c9ede78642f.png)

The test circuit continued to work with no pullup at 921600bps, but the rise time was clearly too low, leaving no margin for error:

![ch340e-74x125-1k-sig-xR](https://user-images.githubusercontent.com/236050/148160471-a4790658-f4f6-4caf-b466-d7d6932b4e33.png)

Of the adapters I tested, the CH340E had the strongest internal RX pullup, about 20k on my chip, but an additional pullup would still help to assure 921600bps support.

# Schottky or not?

*Disclaimer: the following is based on my testing and my very basic understanding of the fundamentals. Feedback is welcome.*

Using a Schottky diode instead of a silicon small signal diode between TX and UPDI would affect the circuit in a few ways:

1. UPDI would be driven to a lower low by the adapter's TX pin (perhaps 0.3V instead of 0.6V)
2. Increased reverse leakage current
3. Increased capacitance

Note that even when slowed down by a series resistor, testing has shown that TX can quickly drive UPDI to an acceptably low voltage. The larger concern is the UPDI pin's slow rise time as TX transitions from low to high.

Therefore, for item 1) above, a "lower low" does not provide a real benefit. And, if anything, a lower low may be counter-productive as it may increase the time it takes the UPDI pin to rise to an acceptable high voltage level.

Item 2) actually helps. Given that some sort of pullup is necessary on the UPDI pin, leakage current is beneficial. However, a parallel resistor achieves the same in a much more straightforward and controllable fashion.

Item 3) isn't a huge concern in the test circuit, but may cause problems for real-world UPDI adapters as using a Schottky diode causes slower rise and fall times as the UPDI pin asserts high and low.

Tests were performed with a big, slow [1N5817 Schottkt Diode](https://www.diodes.com/assets/Datasheets/ds23001.pdf) in place of the 1N4148 signal diode, with interesting results. The parallel 4.7k resistor was also removed. The CH340E adapter was used.

First, we can see that despite the removal of the parallel resistor, rise times on the UPDI pin when driven by TX were very fast:

![ch340e-74x125-1k-schottky-xR-1](https://user-images.githubusercontent.com/236050/148160509-d140c897-e637-43fc-b8ac-0ade060839ae.png)
*CH340E with Schottky and without Parallel Resistor*

Second, we can see the drag on rise and fall times when driven by the UPDI pin:

![ch340e-74x125-1k-schottky-xR-2](https://user-images.githubusercontent.com/236050/148160568-461339ee-6c42-4829-92fe-344bcaba00d7.png)
*CH340E with Schottky and without Parallel Resistor*

In my estimation, using a Schottky diode increases complexity without significant benefit, and may cause problems when the UPDI pin drives the line.

# Conclusions

- For most serial adapters, it looks like 921600bps will be very achievable, even on a sloppy breadboard circuit (still need to try out circuits without the buffers...)
- A parallel resistor should be considered for higher data rates unless the serial adapter has a very strong internal RX pullup, and even then, it's probably a good idea
- Surprisingly, using a Schottky diode has a bigger positive impact on rise times than fall times
- A Schottky diode may not be preferable to a silicon small signal diode and a properly sized parallel resistor. Simplicity rules.
- Adapter specific settings may need to be provided to `prog.py` for higher data rates. If rise and fall times are within 1µs, failures may be on the software side.
- 115200bps "always" works. Above that, results may vary based on hardware and software
- Due to driver differences, adapter performance will vary across computing platforms

# Corrections

- I originally reported testing with an FTDI cable. It was actually a CP2102 adapter.

**Comment by hmeijdam** [comment]
Compliments on your interesting post.

**Comment by SpenceKonde** [comment]
Did you find that the write delay resulted in actually "turning a profit" in terms of time? my experience was that the actual total upload time suffered significantly with the write delay, because the python sleep() function sleeps for longer than requested. Upload time is often dominated by USB latency rather than transfer time. Also with write delay of 1, you will likely get the same behavior as 2, because it actually sleeps for significantly longer than that. 

There is the implied buffer of 3 bytes (the TXDATAL, the buffer behind it, and the shift register, that can be filled, and data is lost at the start bit of the fourth byte. I am not very good at python though, and I couldn't find a way to get tighter control over the timing in python. That the difference between 115200 baud and 921600 baud is enough to make it work or not on so many parts without write delay is amazing. that's 36 * ((1/115200) - (1/921600) ) = 0.25ms that  makes the difference between success and failure... (36 = 3 bytes * (8 data bits + 1 start bit + 1 parity bit + 2 stop bits))

I have definitely had problems where, without the schottky, the lows didn't get low enough to be recognized, though....

**Reply by jvasileff** [reply]
> Did you find that the write delay resulted in actually "turning a profit"

I recorded times below as reported by `prog.py`, and yes, it appears using `-wd` along with higher bitrates is worth it. It's also interesting that my tests on macOS are much faster than on Linux for the CP2102, and as might be expected as a consequence, failures start occuring at lower bitrates on macOS.

Chip | OS | BAUD | Write Delay | Writing (s) | Verify (s)
--- | --- | --- | --- | --- | ---
CP2102 | Linux | 115200 | -wd 0 | 3.11 | 2.22
CP2102 | Linux | 115200 | -wd 1 | 3.38 | 2.23
CP2102 | Linux | 115200 | -wd 2 | 3.65 | 2.18
CP2102 | Linux | 230400 | -wd 0 | 1.86 | 1.25
CP2102 | Linux | 230400 | -wd 1 | 2.12 | 1.27
CP2102 | Linux | 230400 | -wd 2 | 2.43 | 1.27
CP2102 | Linux | 460800 | -wd 0 | f | f
CP2102 | Linux | 460800 | -wd 1 | 1.57 | 0.93
CP2102 | Linux | 460800 | -wd 2 | 1.83 | 0.94
CP2102 | Linux | 921600 | -wd 0 | f | f
CP2102 | Linux | 921600 | -wd 1 | 1.31 | 0.83
CP2102 | Linux | 921600 | -wd 2 | 1.56 | 0.80

Chip | OS | BAUD | Write Delay | Writing (s) | Verify (s)
--- | --- | --- | --- | --- | ---
CP2102 | MacOS | 115200 | -wd 0 | 2.70 | 1.73
CP2102 | MacOS | 115200 | -wd 1 | 3.04 | 1.73
CP2102 | MacOS | 115200 | -wd 2 | 3.35 | 1.73
CP2102 | MacOS | 230400 | -wd 0 | f | f
CP2102 | MacOS | 230400 | -wd 1 | 1.87 | 0.89
CP2102 | MacOS | 230400 | -wd 2 | 2.18 | 0.89
CP2102 | MacOS | 460800 | -wd 0 | f | f
CP2102 | MacOS | 460800 | -wd 1 | 1.33 | 0.47
CP2102 | MacOS | 460800 | -wd 2 | 1.63 | 0.47
CP2102 | MacOS | 921600 | -wd 0 | f | f
CP2102 | MacOS | 921600 | -wd 1 | f | f
CP2102 | MacOS | 921600 | -wd 2 | 1.39 | 0.26

**Comment by jvasileff** [comment]
> Did you find that the write delay resulted in actually "turning a profit" in terms of time?

I'll test that out later and post some numbers. Although, my main goal is to test the limits so I can find a reasonable compromise between performance and stability.

> Also with write delay of 1, you will likely get the same behavior as 2, because it actually sleeps for significantly longer than that.

Thanks, good to know! I used 2 for "good measure" in case 1 was just borderline-working.

> I have definitely had problems where, without the schottky, the lows didn't get low enough to be recognized, though....

I wonder if there is some variability with the ATtiny chips or generations. I thought I'd try out a "worst case scenario" with the `1626` I've been testing:

- 4.7k TX series resistor
- silicon diode
- 22k RX pullup to VCC (3v3)
- 470Ω series resistor on the UPDI pin

and despite obnoxiously high 1.2V lows, programming was completely reliable at 921600:

![ch340e-4 7k-22k-470r-silicon](https://user-images.githubusercontent.com/236050/148276460-98f1f8d5-1324-4457-9d1c-ae92e4533460.png)

Of course, in this case, it makes sense to replace the diode with a wire, which brings the lows down to 0.8V (and up to 0.5V when the UPDI pin is driving the line):

![ch340e-4 7k-22k-470r-nodiode](https://user-images.githubusercontent.com/236050/148276591-1fc5a64b-0a41-4b41-87d2-7c366263de4a.png)

**Comment by SpenceKonde** [comment]
Edit: OH! you have a buffer in there! Yeah, that makes sense. That's a simple enough addition... 

I'm interested to know if you see the same sort of behavior I did if you don't have the buffer.

**Reply by jvasileff** [reply]
Yeah, with the buffer, the adapters are seeing a clean & strong 3.3V on RX.

But after seeing your message, I double checked, and a big yikes on my end. What I thought was an FTDI cable is actually a CP2102 cable - [this one](https://www.adafruit.com/product/954) specifically. At least my original measurements for my CP2105 and what I now know to be a CP2102 are identical! I'll correct the original post.

I do plan to test without the buffers after I try a few more things with the existing test circuit. My hunch is that the biggest factor for the best diode & resister setup is going to be whether or not a 470Ω (or other) resistor is protecting the UPDI port.

**Comment by SpenceKonde** [comment]
https://stackoverflow.com/questions/43771575/accurate-sleep-delay-within-python-while-loop

Hmmm.... maybe one of those methods could be adapted.

**Comment by apws** [comment]
wow, real science, always like that, tnx :-)

**Comment by SpenceKonde** [comment]
This intrigued me enough to throw together a design based on CH340N and dual buffer, 20mm x 16mm, with a switch to turn power from USB (5v) on or off, 3-pin UPDI connection, with target Vcc always used as the voltage the buffer is powered by, while the CH340N is powered at 5V so it should work across  to fit the holes in the poor-mans-panels I'm going to send out this weekend. A dozen part placements including a PTC fuse so an excessive load and a shoddy USB current limiting circuit (these are common) can't trash the switch (which has low current limit, because it's a tiny SMD switch, and an indicator LED (bicolor hooked up between RTS and TX - so you get an indication if another program opens the port (almost everything drives RTS low, SerialUPDI doesn't, for almost this exact reason - to tell it apart from the majority of programs that might access the port), and an indication when there's transmitting activity. - and it only needs one resistor (trying to keep the board small and partcount low - this is the first time I get to use the 0.1uF capacitor networks I hunted down a few months back: I needed 3 0.1uF caps  right in the middle of the board, each one connected between ground and a different supply rail. For once, those were a useful part.

**Comment by SpenceKonde** [comment]
I just checked in some changes for 2.5.6 that should help with this.

**Reply by jvasileff** [reply]
Thats great, I’ll check it out! From command-line, is the primary change the accuracy of the write delay option, or are there additional changes to test?
